### DISCUSSION TITLE
Structure of the Allegro debugger backend

### DISCUSSION THREAD

**Discussion by rpgoldman** [joaotavora/sly#540] [Q&A]
I would like to make the debugger part of the SLY backend for Allegro more capable.

The debugger for Allegro is one of its best features, but neither SLY nor SLIME makes very effective use of it.

I was trying to add the unimplemented interface `sly-db-break-on-return`.  This seems to require doing `(top-level:do-command :boe)`.  Unfortunately, using a top-level command does not seem to work; even wrapped in `debugger:eval-form-in-context`, I get an error about `Non-structure argument NIL passed to ref of structure slot 14`.

Is it possible to use `top-level:do-command` with the SLY repl?  If not, any idea how to get access to debugger functions?

**Comment by joaotavora** [comment]
Hello @rpgoldman, I'm afraid I'm kind of detached from Common Lisp and Allegro at the moment.  I can tell you that the `sly-db-*` interfaces were added as a carbon copy of SLIME's interfaces and I'm not too familiar with them.

But do try out things and report back here with some ideas or maybe a patch.

If you're interested in an alternative CL portable stepper that is known to work with SLY and Allegro, see https://github.com/joaotavora/sly-stepper (and the supporting ELS [paper](https://zenodo.org/record/3742759), if you're curious)

**Comment by dabrandon** [comment]
已收到！
